### ISSUE TITLE
How to count the number of edges in one module?

### ISSUE THREAD

**Issue by harperchen** [ZHYfeng/Dependency#1]
Hi Yu,

I recently read your paper published at ICSE 2022. Thank you for your talented work. May I know how you count the overall number of edges for the four modules kvm/cdrom/ptmx/snd_seq for Figure 3 in your papers?

Thank you.
Wei

**Comment by ZHYfeng** [comment]
the edge coverage:
https://clang.llvm.org/docs/SanitizerCoverage.html#edge-coverage
by this instrumentation, you can get edge coverage from KCOV.

the overall number of edges:
we have the linked LLVM bitcode of each module.
The total number of basic blocks/edges is the number.
(By using the instrumentation above, it would insert some new basic blocks for edges.)

notice:
using `-fembed-bitcode -save-temps=obj` to make the instrumentation work on not only binary but also the bitcode.

**Comment by harperchen** [comment]
Hi Yu，

Thank you for introducing sanitizer coverage for me. I'm new to this feature. So for the following example,
```
void foo(int *a) {
    if (a) {
        *a = 0;
    }
}
```
As far as I understand, the total number of edges is 4, even if the new block 5 is inserted to deal with the critical edge, right?

```
; Function Attrs: noinline nounwind optnone uwtable
define dso_local void @foo(i32* %0) #0 comdat {
  %2 = alloca i32*, align 8
  call void @__sanitizer_cov_trace_pc_guard(i32* getelementptr inbounds ([4 x i32], [4 x i32]* @__sancov_gen_, i32 0, i32 0)) #2
  store i32* %0, i32** %2, align 8
  %3 = load i32*, i32** %2, align 8
  %4 = icmp ne i32* %3, null
  br i1 %4, label %6, label %5

5:                                                ; preds = %1
  call void @__sanitizer_cov_trace_pc_guard(i32* inttoptr (i64 add (i64 ptrtoint ([4 x i32]* @__sancov_gen_ to i64), i64 4) to i32*)) #2
  br label %8

6:                                                ; preds = %1
  call void @__sanitizer_cov_trace_pc_guard(i32* inttoptr (i64 add (i64 ptrtoint ([4 x i32]* @__sancov_gen_ to i64), i64 8) to i32*)) #2
  %7 = load i32*, i32** %2, align 8
  store i32 0, i32* %7, align 4
  br label %8

8:                                                ; preds = %5, %6
  call void @__sanitizer_cov_trace_pc_guard(i32* inttoptr (i64 add (i64 ptrtoint ([4 x i32]* @__sancov_gen_ to i64), i64 12) to i32*)) #2
  ret void
}
```
It's not easy for me to compile the kernel to LLVM bitcode. Thanks to the virtual machine you provided, I can directly check the LLVM IR of each module. In case I'm correct for the above question, is the following method correct to get the total number of edges of module kvm, say 33781?
![image](https://user-images.githubusercontent.com/31884331/166147586-3f4ed975-df36-4fd2-9808-f6346fa7ba53.png)

Besides, I have two more questions:
1. Have you tried wllvm to compile LLVM bitcode for linux kernel? Does it support coverage sanitizer? Or may I know if the linked [instruction ](https://github.com/ZHYfeng/Build_Linux_Kernel_Into_LLVM_Bitcode/tree/master/01-change_makefile) also works for newer kernel such as linux v5.15 with higher version clang, like clang-12? 
2. How do you identify the necessary source files for each module? For example, for the kvm module, I noticed that except for `virt/kvm` folder, there are some files related to kvm under folder `arch/x86/kvm`, will the latter folder also be counted when generating `built-in.bc` for kvm module?

**Comment by ZHYfeng** [comment]
The edge number you understand is correct.

For the questions:
1. Both wllvm and gllvm are not good enough for the Linux kernel. One important reason is that they can not correctly link the bitcode in the Linux kernel. For the new kernel, you can try this [way](https://github.com/ZHYfeng/Build_Linux_Kernel_Into_LLVM_Bitcode/blob/master/02-replace_cmd_log/README.md), which works for Linux v5.12 and clang-13.
2. Based on the link process in the kernel to find the built-in.bc, which is from the built-in.o when building the kernel. In the KVM case, if you build a kernel for x86, it would link them together.

**Comment by harperchen** [comment]
Hi Yu,

I've tried the way you recommended compiling the kernel to bitcode. I noticed that there are two kinds of coverage in Syzkaller: signal and cover. In case the kernel is compiled by gcc, does cover mean basic block coverage and signal refer to edge coverage? However, for the kernel compiled by clang where `__sanitizer_cov_trace_pc` is instrumented at edge level, is the cover itself directly the number of edges already covered?

Further, if I would like to connect the mapping between binary address such as `0xffffffff84bce074` with the corresponding LLVM IR, what should I do? Do I need to use source line code as the bridge?

Thank you.
